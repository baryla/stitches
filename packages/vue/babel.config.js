module.exports = {
  extends: '../../babel.config.js',
  plugins: ['@vue/babel-plugin-jsx'],
};
