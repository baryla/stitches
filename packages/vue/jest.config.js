const base = require('../../jest.config.base.js');

module.exports = {
  ...base,
  name: '@atomica/css',
  displayName: 'css',
  transform: {
    '\\.ts?x$': ['babel-jest', { cwd: __dirname }],
  },
};
