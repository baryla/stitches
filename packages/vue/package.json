{
  "name": "@stitches/vue",
  "version": "0.0.1",
  "description": "The modern CSS-in-JS library",
  "main": "dist/vue.cjs.js",
  "module": "dist/vue.esm.js",
  "preconstruct": {
    "source": "src/index.tsx"
  },
  "lint-staged": {
    "./**/*.{js,jsx,ts,tsx}": [
      "tslint --fix",
      "prettier --write",
      "git add"
    ],
    "./**/*.{css,md,json}": [
      "prettier --write",
      "git add"
    ]
  },
  "license": "MIT",
  "repository": "git+https://github.com/modulz/stitches.git",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "typecheck": "tsc --noEmit",
    "lint": "tslint --project tsconfig.json --format stylish",
    "format": "prettier '**/*.{md,js,jsx,json,ts,tsx}' --write",
    "test": "jest --env=jsdom --coverage",
    "test:watch": "node ../../node_modules/jest/bin/jest.js --env=jsdom --watch --updateSnapshot",
    "posttest": "npm run typecheck && npm run lint",
    "preversion": "npm test",
    "postversion": "git push && git push --tags"
  },
  "keywords": [
    "typescript",
    "css-in-js",
    "css",
    "library"
  ],
  "files": [
    "dist"
  ],
  "peerDependencies": {
    "vue": "^3.0.0-rc.9"
  },
  "dependencies": {
    "@stitches/core": "^0.0.1",
    "tslib": "^1.11.1"
  },
  "devDependencies": {
    "@types/node": "^13.11.1"
  }
}
