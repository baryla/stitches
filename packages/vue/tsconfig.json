{
  "extends": "../../tsconfig.build.json",
  "include": ["src/**/*", "tests/**/*"],
  "compilerOptions": {
    "jsx": "preserve",
    "preserveSymlinks": true,
    "skipLibCheck": true
  }
}
