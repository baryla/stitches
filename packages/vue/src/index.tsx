import { MAIN_BREAKPOINT_ID, createCss, hashString, TConfig, TCss, TDefaultCss, TMainBreakPoint } from '@stitches/core';
import * as Vue from 'vue';

import type { IntrinsicElements } from '../types/jsx';

export { _ATOM } from '@stitches/core';

const hasWarnedInlineStyle = false;

export type VueProps = Vue.AllowedComponentProps & Vue.VNodeProps & Vue.ComponentCustomProps;
export type TCssProp<T extends TConfig> = TDefaultCss<T> | (string & {});

/**
 * Extracts Variants from an object:
 */
export type TExtractVariants<Styles> = Styles extends {
  variants: infer Variants;
}
  ? { [a in keyof Variants]: keyof Variants[a] }
  : {};

/**
 * Extracts Breakpoint keys from a config
 */
export type BreakPointsKeys<Config extends TConfig> = keyof Config['breakpoints'];

/**
 * Takes a value and if it's one of the string type representations of a boolean ('true' | 'false')
 * it adds the actual boolean values to it
 */
export type CastStringToBoolean<Val> = Val extends 'true' | 'false' ? boolean | 'true' | 'false' : never;

/**
 * Takes a variants object and converts it to the correct type information for usage in props
 */
export type VariantASProps<Config extends TConfig, VariantsObj> = {
  [V in keyof VariantsObj]?:
    | CastStringToBoolean<VariantsObj[V]>
    | VariantsObj[V]
    | {
        [B in BreakPointsKeys<Config> | TMainBreakPoint]?: CastStringToBoolean<VariantsObj[V]> | VariantsObj[V];
      };
};

/**
 * Types for a styled component which contain:
 * 1. First overload which matches when the as prop isn't passed
 * 2. Second overload which matches when the as prop *IS* passed
 * 3. The compoundVariants function typings
 */
export interface IStyledComponent<
  ComponentOrTag extends keyof IntrinsicElements | Vue.Component,
  Variants,
  Config extends TConfig
> {
  /**
   * First overload without the "as" prop
   */
  (
    props: VueProps & {
      as?: never;
      css?: TCssWithBreakpoints<Config>;
      className?: string;
      children?: any;
    } & VariantASProps<Config, Variants>
  ): any;
  /**
   * Second overload * WITH * the "as" prop.
   */
  <AS extends keyof IntrinsicElements | Vue.Component>(
    props: {
      as: AS;
      css?: TCssWithBreakpoints<Config>;
      className?: string;
      children?: any;
    } & VariantASProps<Config, Variants> &
      VueProps
  ): any;
  /**
   * Compound Variant typing:
   */
  compoundVariant: (
    compoundVariants: VariantASProps<Config, Variants>,
    possibleValues: TCssWithBreakpoints<Config>
  ) => IStyledComponent<ComponentOrTag, Variants, Config>;
  /**
   * Default props typing:
   */
  defaultProps?: VariantASProps<Config, Variants> & { [k: string]: any };
  /**
   * DisplayName typing:
   */
  displayName?: string;
}

/** Typed css with tokens and breakpoints */
export type TCssWithBreakpoints<Config extends TConfig> = TCssProp<Config> &
  { [key in BreakPointsKeys<Config>]?: TCssProp<Config> };

/** The type for the styles in a styled call */
export type TComponentStylesObject<Config extends TConfig> = TCssWithBreakpoints<Config> & {
  variants?: {
    [k: string]: {
      [s: string]: TCssWithBreakpoints<Config>;
    };
  };
};
/**
 * Types for styled.button, styled.div, etc..
 */
export type TProxyStyledElements<Config extends TConfig> = {
  [key in keyof IntrinsicElements]: <BaseAndVariantStyles extends TComponentStylesObject<Config>>(
    a: BaseAndVariantStyles | TComponentStylesObject<Config>
  ) => IStyledComponent<key, TExtractVariants<BaseAndVariantStyles>, Config>;
};
/**
 * Styled Components creator Type.
 * ie: styled.div(styles) | styled('div', {styles})
 */
export type TStyled<Config extends TConfig> = {
  // tslint:disable-next-line: callable-types
  <
    TagOrComponent extends keyof IntrinsicElements | IStyledComponent<any, any, Config>,
    BaseAndVariantStyles extends TComponentStylesObject<Config>
  >(
    tag: TagOrComponent,
    baseStyles: BaseAndVariantStyles | TComponentStylesObject<Config>
  ): TagOrComponent extends IStyledComponent<any, any, Config>
    ? TagOrComponent
    : IStyledComponent<TagOrComponent, TExtractVariants<BaseAndVariantStyles>, Config>;
} & TProxyStyledElements<Config>;

const createCompoundVariantsMatcher = (breakPoints: any, existingMap?: any) => {
  const map = new Map();
  map.set(MAIN_BREAKPOINT_ID, [...(existingMap?.get(MAIN_BREAKPOINT_ID) || [])]);
  Object.keys(breakPoints).forEach((breakpoint) => map.set(breakpoint, [...(existingMap?.get(breakpoint) || [])]));
  return map;
};

const getChildren = (slots: Vue.Slots) => (slots.default ? slots.default() : []);

export const createStyled = <Config extends TConfig>(
  config: Config
): {
  css: TCss<Config>;
  styled: TStyled<Config>;
} => {
  const css = createCss(config);
  const defaultElement = 'div';

  const Box = ({ as, ...restProps }: any, ctx: Vue.SetupContext) =>
    Vue.createVNode(as || defaultElement, restProps, getChildren(ctx.slots));

  let currentAs: string | undefined;

  const configBreakpoints = config.breakpoints || {};

  const styledInstance = (
    baseAndVariantStyles: any = (cssComposer: any) => cssComposer.compose(),
    Component: Vue.Component = Box
  ) => {
    let numberOfCompoundVariants = 0;
    const as = currentAs;
    const { variants = {}, ...base } = baseAndVariantStyles;
    const baseStyles: any = css(base);
    // compound s vars & constants:
    // keep track of all compound variants:
    const compoundVariants: any[] = [];
    // a map that keeps track of the required number of matching s left for each break point:
    const requiredMatches = createCompoundVariantsMatcher(configBreakpoints);
    // keep track of the number of available variants
    const evaluatedVariantMap: Map<string, Map<string, { [key: string]: string }>> = new Map();
    // store pre evaluated variants
    const evaluatedCompoundVariants: Map<any, { [key: string]: string }> = new Map();

    // tslint:disable-next-line: forin
    for (const Name in variants) {
      const variantMap: Map<string, { [key: string]: string }> = new Map();
      // tslint:disable-next-line: forin
      for (const ValueName in variants[Name]) {
        const evaluatedStyles = evaluateStylesForAllBreakpoints(variants[Name][ValueName], configBreakpoints, css);
        variantMap.set(ValueName, evaluatedStyles);
      }
      evaluatedVariantMap.set(Name, variantMap);
    }

    const stitchesComponentId = `scid-${hashString(JSON.stringify(baseAndVariantStyles))}`;

    const StitchesComponent = (props: any, ctx: Vue.SetupContext) => {
      // if (process.env.NODE_ENV === 'development') {
      //   // we're breaking the rules of hooks on purpose as the env will never change
      //   // eslint-disable-next-line
      //   const memoStyled = React.useMemo(() => props.css, []); // We want this to only eval once
      //   if (memoStyled !== props.css && !hasWarnedInlineStyle) {
      //     // tslint:disable-next-line
      //     console.warn(
      //       '@stitches/react : The css prop should ideally not be dynamic. Define it outside your component using the css composer, or use a memo hook'
      //     );
      //     hasWarnedInlineStyle = true;
      //   }
      // }

      const compositions = [baseStyles];

      const propsWithoutVariantsAndCssProp: any = {};
      // clone the compound s matcher
      const compoundRequiredMatches = createCompoundVariantsMatcher(configBreakpoints, requiredMatches);
      // keep track of the number of unResolved s so that we could bail early:
      const numberOfUnResolvedCompoundVariants = {
        current: numberOfCompoundVariants,
      };
      for (const key in props) {
        // check if the prop is a variant
        if (key in variants) {
          const evaluatedVariant = evaluatedVariantMap.get(key);
          // normalize the value so that we only have to deal with one structure:
          const keyVal =
            props[key] && typeof props[key] !== 'object' ? { [MAIN_BREAKPOINT_ID]: props[key] } : props[key];
          // tslint:disable-next-line: forin
          for (const breakpoint in keyVal) {
            // check if the variant exist for this breakpoint
            if (keyVal[breakpoint] && evaluatedVariant && evaluatedVariant.get(String(keyVal[breakpoint]))) {
              compositions.push(evaluatedVariant.get(String(keyVal[breakpoint]))?.[breakpoint]);
            }
            /** Compound variants: */
            if (numberOfUnResolvedCompoundVariants.current) {
              compoundVariants.forEach((compoundVariant, i) => {
                // if this breakpoint  matches a compound
                // eslint-disable-next-line
                if (String(keyVal[breakpoint]) === String(compoundVariant[key])) {
                  compoundRequiredMatches.get(breakpoint)[i]--;
                }
                // when the required matches reach 0 for any compound ...
                // we know we have a matched compoundVariant
                if (compoundRequiredMatches.get(breakpoint)[i] === 0) {
                  numberOfUnResolvedCompoundVariants.current--;
                  compositions.push(evaluatedCompoundVariants.get(compoundVariant)?.[breakpoint]);
                }
              });
            }
            /** End compound variants */
          }
        } else {
          propsWithoutVariantsAndCssProp[key] = props[key];
        }
      }

      if (propsWithoutVariantsAndCssProp.css) {
        compositions.push(propsWithoutVariantsAndCssProp.css);
        propsWithoutVariantsAndCssProp.css = undefined;
      }

      return Vue.createVNode(
        Component,
        {
          ...propsWithoutVariantsAndCssProp,
          as: props.as || as,
          className: css(stitchesComponentId, ...compositions, props.className),
        },
        getChildren(ctx.slots)
      );
    };

    StitchesComponent.toString = () => `.${stitchesComponentId}`;

    (StitchesComponent as any).compoundVariant = (compundVariantsObject: any, compoundVariantStyles: any) => {
      // Update component level variables:
      numberOfCompoundVariants++;
      // Each time we add
      compoundVariants.push(compundVariantsObject);
      // required matches is a map with breakpoints
      // each time we add a compound variant, we also push its length to
      // all of the breakpoints so:
      // requiredMatches.get(breakpoint)[i] === Object.keys(compoundVariants[i]).length
      // at render time we clone the requiredMatches map and whenever a prop matches a compound variant we decrement
      // the required matches for this compound variant at this breakpoint
      // when the required matches hit 0 we know it's a mtach
      requiredMatches.forEach((value, key) => {
        value.push(Object.keys(compundVariantsObject).length);
      });

      const evaluatedStyles = evaluateStylesForAllBreakpoints(compoundVariantStyles, configBreakpoints, css);

      evaluatedCompoundVariants.set(compundVariantsObject, evaluatedStyles);
      return StitchesComponent;
    };
    return StitchesComponent;
  };

  // tslint:disable-next-line
  const styledProxy = new Proxy(() => {}, {
    get(_, prop) {
      if (prop === 'Box') {
        return Box;
      }
      currentAs = String(prop);
      return styledInstance;
    },
    apply(_, __, [Element, styling]) {
      if (typeof Element === 'string') {
        currentAs = Element;
        return styledInstance(styling);
      }
      currentAs = undefined;
      return styledInstance(styling, Element);
    },
  });

  return {
    styled: styledProxy as any,
    css,
  };
};

function evaluateStylesForAllBreakpoints(styleObject: any, configBreakpoints: any, css: any) {
  const breakpoints: { [key: string]: string } = {
    [MAIN_BREAKPOINT_ID]: css(styleObject),
  };
  if (configBreakpoints) {
    // tslint:disable-next-line
    for (const breakpoint in configBreakpoints) {
      breakpoints[breakpoint] = css({
        [breakpoint]: styleObject,
      });
    }
  }
  return breakpoints;
}
