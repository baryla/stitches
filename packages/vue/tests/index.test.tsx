import * as Vue from 'vue';
import { mount } from '@vue/test-utils';
import { _ATOM } from '@stitches/core';
import { createStyled } from '../src';

let { styled, css } = createStyled({});

beforeEach(() => {
  const newStyled = createStyled({});
  styled = newStyled.styled;
  css = newStyled.css;
});

afterEach(() => {
  (css as any).dispose();
});

describe('styled - vue', () => {
  test('Renders basic component', () => {
    const Heading = styled('h1', {
      color: 'red',
    });

    const App = () => <Heading>Hello World!</Heading>;

    const mountedApp = mount(App);

    expect(mountedApp.html()).toMatchSnapshot();
  });

  test('renders component with variant', () => {
    const Button = styled('button', {
      backgroundColor: 'gray',
      variants: {
        variant: {
          red: { backgroundColor: 'red' },
          blue: { backgroundColor: 'blue' },
        },
      },
    });

    const App = () => (
      <div>
        <Button>Hello World!</Button>
        <Button variant="red">Hello World!</Button>
        <Button variant="blue">Hello World!</Button>
      </div>
    );

    const mountedApp = mount(App);

    expect(mountedApp.html()).toMatchSnapshot();
  });

  test('Handles css prop', () => {
    const Button = styled('button', {
      backgroundColor: 'gray',
    });

    const App = () => <Button css={{ color: 'red' }}>Hello World</Button>;

    const mountedApp = mount(App);

    expect(mountedApp.html()).toMatchSnapshot();
  });

  test('Handles compound variant', () => {
    const Button = styled('button', {
      backgroundColor: 'gray',
      variants: {
        variant: {
          red: {
            backgroundColor: 'red',
          },
        },
        size: {
          1: {
            height: 30,
          },
          2: {
            height: 60,
          },
        },
      },
    }).compoundVariant(
      {
        variant: 'red',
        size: 1,
      },
      {
        backgroundColor: 'compoundColor',
      }
    );

    const App = () => (
      <>
        <Button variant="red" size={2}>
          not compound
        </Button>
        <Button variant="red" size={1}>
          compound
        </Button>
      </>
    );

    const mountedApp = mount(App);

    expect(mountedApp.html()).toMatchSnapshot();
  });

  test('It has default displayName when a string based element is passed', () => {
    const Button = styled('button', {});
    expect(Button.displayName).toBe('styled(button)');
  });

  test('It defaults to styled(component) when styling a component that has no displayName', () => {
    const Component = () => <div></div>;
    const Button = styled(Component, {});
    expect(Button.displayName).toBe('styled(Component)');
  });

  test("It defaults to a displayName that uses the internal styled component's displayName if it has one", () => {
    const Component = () => <div></div>;
    Component.displayName = 'MyFancyComponent';
    const Button = styled(Component, {});
    expect(Button.displayName).toBe(`styled(${Component.displayName})`);
  });

  test('Breakpoints work in variants', () => {
    const { styled: _styled, css: _css } = createStyled({
      showFriendlyClassnames: true,
      breakpoints: {
        breakpointOne: (rule) => `@media(min-width:400px){${rule}}`,
        breakpointTwo: (rule) => `@media(min-width:800px){${rule}}`,
      },
    });
    const Button = _styled('button', {
      color: 'red',
      breakpointOne: {
        height: '10px',
      },
      breakpointTwo: {
        height: '20px',
      },
      variants: {
        size: {
          big: {
            breakpointOne: {
              height: '100px',
            },
          },
        },
      },
    });
    const App = () => <Button>no variant</Button>;
    const mountedApp = mount(App);

    expect(mountedApp.html()).toMatchSnapshot();

    (_css as any).dispose();
  });

  test('Breakpoints work in variants and when the variant is passed to the jsx element', () => {
    const { styled: _styled, css: _css } = createStyled({
      showFriendlyClassnames: true,
      breakpoints: {
        breakpointOne: (rule) => `@media(min-width:400px){${rule}}`,
        breakpointTwo: (rule) => `@media(min-width:800px){${rule}}`,
      },
    });
    const Button = _styled('button', {
      color: 'red',
      breakpointOne: {
        height: '10px',
      },
      breakpointTwo: {
        height: '20px',
      },
      variants: {
        size: {
          small: {
            breakpointOne: {
              height: '100px',
            },
            breakpointTwo: {
              height: '200px',
            },
          },
          big: {
            breakpointOne: {
              height: '1000px',
            },
            breakpointTwo: {
              height: '2000px',
            },
          },
        },
      },
    });
    const App = () => <Button size="big">with variant</Button>;
    const mountedApp = mount(App);

    expect(mountedApp.html()).toMatchSnapshot();

    (_css as any).dispose();
  });

  test('Breakpoints work in variants and when a responsive variant is passed ot the element', () => {
    const { styled: _styled, css: _css } = createStyled({
      showFriendlyClassnames: true,
      breakpoints: {
        breakpointOne: (rule) => `@media(min-width:400px){${rule}}`,
        breakpointTwo: (rule) => `@media(min-width:800px){${rule}}`,
      },
    });
    const Button = _styled('button', {
      color: 'red',
      breakpointOne: {
        height: '10px',
      },
      breakpointTwo: {
        height: '20px',
      },
      variants: {
        size: {
          small: {
            breakpointOne: {
              height: '100px',
            },
            breakpointTwo: {
              height: '200px',
            },
          },
          big: {
            breakpointOne: {
              height: '1000px',
            },
            breakpointTwo: {
              height: '2000px',
            },
          },
        },
      },
    });
    const App = () => <Button size={{ breakpointOne: 'small' }}>with responsive variant</Button>;
    const mountedApp = mount(App);

    expect(mountedApp.html()).toMatchSnapshot();
    (_css as any).dispose();
  });
});
