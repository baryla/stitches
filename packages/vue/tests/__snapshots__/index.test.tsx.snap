// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styled - vue Breakpoints work in variants 1`] = `"<button class=\\"_initial_c_dzoaVP _breakpointOne_h_hGDbbG _breakpointTwo_h_cvoKvR scid-kvQUME\\">no variant</button>"`;

exports[`styled - vue Breakpoints work in variants and when a responsive variant is passed ot the element 1`] = `"<button class=\\"_breakpointOne_h_ctNPWK _breakpointTwo_h_clZhDR _initial_c_dzoaVP scid-cXpAaU\\">with responsive variant</button>"`;

exports[`styled - vue Breakpoints work in variants and when the variant is passed to the jsx element 1`] = `"<button class=\\"_breakpointOne_h_inlnue _breakpointTwo_h_dbWWOp _initial_c_dzoaVP scid-cXpAaU\\">with variant</button>"`;

exports[`styled - vue Handles compound variant 1`] = `"<button class=\\"_dtTvho _bieopk scid-URbJg\\">not compound</button><button class=\\"_hEGtPv _bSHlld scid-URbJg\\">compound</button>"`;

exports[`styled - vue Handles css prop 1`] = `"<button class=\\"_dzoaVP _iBRUhC scid-dmZJpl\\">Hello World</button>"`;

exports[`styled - vue Renders basic component 1`] = `"<h1 class=\\"_dzoaVP scid-gmqXFB\\">Hello World!</h1>"`;

exports[`styled - vue renders component with variant 1`] = `"<div><button class=\\"_iBRUhC scid-gFxMiu\\">Hello World!</button><button class=\\"_bieopk scid-gFxMiu\\">Hello World!</button><button class=\\"_YfjLh scid-gFxMiu\\">Hello World!</button></div>"`;
